cli/login/login.go

```diff
@@ -1,33 +1,98 @@
 package login
 
 import (
+	"context"
+	"flag"
 	"fmt"
 	"io"
 	"os"
 	"os/exec"
 	"runtime"
 	"strings"
+	"sync"
 
 	"github.com/buildbuddy-io/buildbuddy/cli/arg"
 	"github.com/buildbuddy-io/buildbuddy/cli/log"
 	"github.com/buildbuddy-io/buildbuddy/cli/parser"
 	"github.com/buildbuddy-io/buildbuddy/cli/storage"
+	"github.com/buildbuddy-io/buildbuddy/server/util/authutil"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"google.golang.org/grpc/metadata"
+
+	bbspb "github.com/buildbuddy-io/buildbuddy/proto/buildbuddy_service"
+	uspb "github.com/buildbuddy-io/buildbuddy/proto/user"
 )
 
 const (
-	apiKeyRepoSetting = "api-key"
+	ApiKeyRepoSetting = "api-key"
 	apiKeyHeader      = "remote_header=x-buildbuddy-api-key"
-	loginURL          = "https://app.buildbuddy.io/settings/cli-login"
 )
 
+var (
+	loginFlagSet = flag.NewFlagSet("login", flag.ContinueOnError)
+
+	apiTarget = loginFlagSet.String("target", "grpcs://remote.buildbuddy.io", "BuildBuddy gRPC target")
+	loginURL  = loginFlagSet.String("login_url", "https://app.buildbuddy.io/settings/cli-login", "URL for user to login")
+)
+
+var getApiClient = sync.OnceValues(func() (bbspb.BuildBuddyServiceClient, error) {
+	conn, err := grpc_client.DialSimple(*apiTarget)
+	if err != nil {
+		return nil, fmt.Errorf("unable to dial BuildBuddy target: %v", err)
+	}
+	return bbspb.NewBuildBuddyServiceClient(conn), nil
+})
+
+func isValidApiKey(apiKey string) bool {
+	bbsClient, err := getApiClient()
+	if err != nil {
+		log.Debugf("unable to get buildbuddy service client: %v", err)
+		return false
+	}
+	ctx := context.Background()
+	ctx = metadata.AppendToOutgoingContext(ctx, "x-buildbuddy-api-key", apiKey)
+
+	_, err = bbsClient.GetUser(ctx, &uspb.GetUserRequest{})
+	if err != nil && status.IsNotFoundError(err) && strings.Contains(err.Error(), authutil.UserNotFoundMsg) {
+		// Org-level API key is used
+		return true
+	}
+	if err != nil {
+		log.Debugf("fail to get user: %v", err)
+		return false
+	}
+	return true
+}
+
+func hasValidApiKey() bool {
+	apiKey, err := storage.ReadRepoConfig(ApiKeyRepoSetting)
+	if err != nil {
+		log.Debugf("unable to read git repo config: %v", err)
+		return false
+	}
+	log.Printf("Found an API key in your .git/config. Verifying it's validity.")
+
+	return isValidApiKey(apiKey)
+}
+
 func HandleLogin(args []string) (exitCode int, err error) {
-	if err := openInBrowser(loginURL); err != nil {
+	if hasValidApiKey() {
+		log.Printf("Current API key is valid. Skipping login.")
+		return 0, nil
+	}
```

Comment:
@brentleyjones I added a `--check` flag. Basic usage would look like this:

```
bb login --check || bb login
```

Slightly more advanced usage if you want control over the specific failure conditions:

```
if ! bb login --check ; then
  exit_code=$?
  if [[ "$exit_code" == 2 ]]; then
    bb login
  else
    exit "$exit_code" # network error - don't bother running `bb login`
  fi
fi
```

Reply:
Actually, to make life easy, I added an `--allow_existing` option which basically does the same logic as that bash script above. So you can just call `bb login --allow_existing` in your scripts, and it will silently succeed if the user is logged in, otherwise prompt them for login.